qiskit/algorithms/gradients/base_sampler_gradient.py

```diff
@@ -91,17 +114,121 @@ def _run(
+        """Preprocess the gradient. This makes a gradient circuit for each circuit. The gradient
+        circuit is a transpiled circuit by using the supported gates, and has unique parameters.
+        ``parameter_values`` and ``parameters`` are also updated to match the gradient circuit.
+
+        Args:
+            circuits: The list of quantum circuits to compute the gradients.
+            parameter_values: The list of parameter values to be bound to the circuit.
+            parameters: The sequence of parameters to calculate only the gradients of the specified
+                parameters.
+            supported_gates: The supported gates used to transpile the circuit.
+
+        Returns:
+            The list of gradient circuits, the list of parameter values, and the list of parameters.
+            parameter_values and parameters are updated to match the gradient circuit.
+        """
+        g_circuits, g_parameter_values, g_parameter_sets = [], [], []
+        for circuit, parameter_value_, parameter_set in zip(
+            circuits, parameter_values, parameter_sets
+        ):
+            circuit_key = _circuit_key(circuit)
+            if circuit_key not in self._gradient_circuit_cache:
+                transpiled_circuit = transpile(
+                    circuit, basis_gates=supported_gates, optimization_level=0
+                )
+                self._gradient_circuit_cache[circuit_key] = _assign_unique_parameters(
+                    transpiled_circuit
+                )
+            gradient_circuit = self._gradient_circuit_cache[circuit_key]
+            g_circuits.append(gradient_circuit.gradient_circuit)
+            g_parameter_values.append(
+                _make_gradient_parameter_values(circuit, gradient_circuit, parameter_value_)
+            )
+            g_parameter_sets.append(_make_gradient_parameter_set(gradient_circuit, parameter_set))
+        return g_circuits, g_parameter_values, g_parameter_sets
+
+    def _postprocess(
+        self,
+        results: SamplerGradientResult,
+        circuits: Sequence[QuantumCircuit],
+        parameter_values: Sequence[Sequence[float]],
+        parameter_sets: Sequence[set[Parameter] | None],
+    ) -> SamplerGradientResult:
+        """Postprocess the gradient. This computes the gradient of the original circuit from the
+        gradient of the gradient circuit by using the chain rule.
+
+        Args:
+            results: The results of the gradient of the gradient circuits.
+            circuits: The list of quantum circuits to compute the gradients.
+            parameter_values: The list of parameter values to be bound to the circuit.
+            parameters: The sequence of parameters to calculate only the gradients of the specified
+                parameters.
+
+        Returns:
+            The results of the gradient of the original circuits.
+        """
+        gradients, metadata = [], []
```

Comment:
This code is mostly the same as in the estimator, up to the `grad_dist` construction. Is there a way we can avoid the code duplication here? 🤔

Reply:
We can move `_preprocess()` and `_postprcess()` in the base sampler gradient and base estimator gradient to `util.py`. Then, we can add some optional argument like `estimator_grad=True` to chose which type of results are expected.